### ISSUE TITLE
`improper_ctypes` lint triggers on valid `#[repr(transparent)]` types

### ISSUE THREAD

**Issue by madsmtm** [rust-lang/rust#115457]
### Code

```Rust
#[repr(C)]
pub struct Foo {
    inner: [u8; 0]
}

extern "C" {
    // Lint does not trigger, as expected
    pub fn foo(_: &Foo);
}

#[repr(transparent)]
pub struct Bar(Foo);

extern "C" {
    // Lint triggers unexpectedly
    pub fn bar(_: &Bar);
}
```

### Current output

```Shell
warning: `extern` block uses type `Bar`, which is not FFI-safe
  --> <source>:16:19
   |
16 |     pub fn bar(_: &Bar);
   |                   ^^^^ not FFI-safe
   |
   = note: this struct contains only zero-sized fields
note: the type is defined here
  --> <source>:12:1
   |
12 | pub struct Bar(Foo);
   | ^^^^^^^^^^^^^^
   = note: `#[warn(improper_ctypes)]` on by default

warning: 1 warning emitted
```

### Desired output

```Shell
The lint should not fire
```

### Rationale and extra context

Reproducible from Rust 1.57.0 (where `#[repr(transparent)]` on a zero-sized struct was made valid) to the current `nightly-2023-08-31`

**Comment by asquared31415** [comment]
I find it odd that this *doesn't* lint on `Foo`, it contains only zero sized fields too.

**Comment by madsmtm** [comment]
If it's an easier pill to swallow, it also lints on `*const Bar` but not on `*const Foo`.

Note that using zero-sized types to represent opaque/extern types like this is a [documented pattern](https://doc.rust-lang.org/nomicon/ffi.html#representing-opaque-structs), so the lint should definitely not fire there.

**Comment by madsmtm** [comment]
Actually, I just saw [this Pre-RFC about it](https://internals.rust-lang.org/t/pre-rfc-define-the-behavior-of-repr-transparent-when-all-fields-are-zero-sized-types/19415), so maybe this pattern was not actually fully specified?

**Comment by surechen** [comment]
> Actually, I just saw [this Pre-RFC about `#[repr(transparent)]` to ZSTs](https://internals.rust-lang.org/t/pre-rfc-define-the-behavior-of-repr-transparent-when-all-fields-are-zero-sized-types/19415), so maybe this pattern was not actually fully specified?
Hi, Thanks very much for finding the informations. So the current errors reported here is consistent with [the reference](https://doc.rust-lang.org/reference/type-layout.html#the-transparent-representation:~:text=is%20a%20bug.-,The%20transparent%20Representation,-The%20transparent%20representation) ?

**Comment by surechen** [comment]
I find that ignoring the warning conditions for the example seems easy to implement, but there doesn't seem to be a definite conclusion here. I think we should continue the discussion in this issue [rust-lang/rust#100954](https://github.com/rust-lang/rust/issues/100954)

**Comment by workingjubilee** [comment]
nb: It also fails to trigger on invalid `repr(transparent)` types:
- https://github.com/rust-lang/rust/issues/116959
